src/DomainHost.sol

```diff
@@ -0,0 +1,490 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.15;
+
+import "./TeleportGUID.sol";
+
+interface DomainGuestLike {
+    function deposit(address to, uint256 amount) external;
+    function lift(uint256 _lid, uint256 wad) external;
+    function rectify(uint256 _lid, uint256 wad) external;
+    function cage(uint256 _lid) external;
+    function exit(address usr, uint256 wad) external;
+    function finalizeRegisterMint(TeleportGUID calldata teleport) external;
+    function finalizeSettle(bytes32 sourceDomain, bytes32 targetDomain, uint256 amount) external;
+}
+
+interface VatLike {
+    function live() external view returns (uint256);
+    function hope(address usr) external;
+    function slip(bytes32 ilk, address usr, int256 wad) external;
+    function frob(bytes32 i, address u, address v, address w, int dink, int dart) external;
+    function suck(address u, address v, uint256 rad) external;
+    function urns(bytes32, address) external view returns (uint256, uint256);
+    function grab(bytes32, address, address, address, int256, int256) external;
+}
+
+interface DaiJoinLike {
+    function vat() external view returns (VatLike);
+    function dai() external view returns (DaiLike);
+    function join(address usr, uint256 wad) external;
+    function exit(address usr, uint256 wad) external;
+}
+
+interface DaiLike {
+    function transfer(address dst, uint256 wad) external returns (bool);
+    function transferFrom(address src, address dst, uint256 wad) external returns (bool);
+    function approve(address usr, uint wad) external returns (bool);
+}
+
+interface RouterLike {
+    function settle(bytes32 sourceDomain, bytes32 targetDomain, uint256 amount) external;
+    function registerMint(TeleportGUID calldata teleport) external;
+}
+
+/// @title Support for xchain MCD, canonical DAI and Maker Teleport - host instance
+/// @dev This is just the business logic which needs concrete message-passing implementation
+abstract contract DomainHost {
+
+    // --- Data ---
+    mapping (address => uint256) public wards;
+    mapping (bytes32 => bool)    public teleports;
+    mapping (bytes32 => mapping (bytes32 => uint256)) public settlements;
+
+    address public vow;
+    uint256 public lid;         // Local ordering id
+    uint256 public rid;         // Remote ordering id
+    uint256 public line;        // Remote domain global debt ceiling [RAD]
+    uint256 public grain;       // Keep track of the pre-minted DAI in the escrow [WAD]
```

Comment:
Is `grain` necessary as a separate storage var? Except during gs it is equal to the amount of DAI in the domain escrow. During gs it is `debt / grain` that is actually used not the `grain` it self